sycl/source/platform.cpp

```diff
@@ -56,19 +56,35 @@ std::vector<platform> platform::get_platforms() {
 
 backend platform::get_backend() const noexcept { return impl->getBackend(); }
 
+#ifdef __INTEL_PREVIEW_BREAKING_CHANGES
+template <typename Param>
+typename detail::ABINeutralT_t<
+    typename detail::is_platform_info_desc<Param>::return_type>
+platform::get_info_impl() const {
+  auto Info = impl->template get_info<Param>();
+  return detail::convert_to_abi_neutral(Info);
```

Comment:
Looking into this example, once can also use `#if` to define these convert helpers and `ABINeutralT_t` trait in such a way that this client code would be uniform between preview/non-preview.

Reply:
Please note that we need to leave the old ABI of `get_info<>` as is.
The old ABI can't handle polymorphic return values depending on the template parameter <T>.
That's why we needed `get_info_impl<>` to return a polymorphic return value to get around the ABI incompatibility issues.
That means we can't avoid using `#ifdef __INTEL_PREVIEW_BREAKING_CHANGES`.

Reply:
Why can't we have

```
#ifndef __INTEL_PREVIEW_BREAKING_CHANGES
template <typename T>
auto convert_to/from_abi_neutral(T &&x) { return std::forward<T>(x); }
#endif
```

Reply:
You mean you want to call this function from the non-preview mode get_inf<>?
It seems like you have an idea how to combine preview and non-preview mode of `get_info`. Correct me if I misunderstood you.

`conver_to/from_abi_neutral<>`  are called from `get_info_impl<>`.
So, we already can't remove `#ifdef __INTEL_PREVIEW_BREAKING_CHANGES` between `get_info<>` and `get_info_impl<>`.  The current ABI of `get_info<>` should remain the same per spec.
Sorry, but I can't imagine the point of adding that code for the `get_info<>` non-preview mode since we can't combine the caller side(`get_info` and `get_info_impl`) unconditionally to `ABINeutralT_t get_info<>.`

Reply:
Let's talk about this particular piece of code. If we you implemented the change above, could we change line 71 to be the same as lines 64/65 (potentially merged into a single statement)? If not, why? I don't see any reason why it wouldn't be possible. If yes, what other places could be simplified the same way?

Reply:
Probably I am missing something. Please fill the gap if I missed anything.
No because the return value is different.
The original get_info() returns std::string, but get_info_impl() returns detail::string for some template parameters.
And we can't change the return type of the original get_info() since it is written in spec.
Can we change it without changing the spec?

Reply:
`auto convert_to/from_abi_neutral(T &&x) { return std::forward<T>(x); }` is a no-op. It doesn't change any type.

Reply:
Correct.
But, I still do not understand how we can copy line 64-65 and replace line 71 without changing the return value of get_info<>.

Reply:
Why doesn't simple 

```
  return detail::convert_to_abi_neutral(impl->template get_info<Param>());
```

work?

Reply:
It seems you are still claiming that the return type of get_info<> is unchanged. Right?
That part I don't understand. So, please help me understand.

The return type of that call depends on` #ifdef __INTEL_PREVIEW_BREAKING_CHANGES`.
It will return a `detail::string` or a `std::strin`g.
The current return type of `get_info<>` is not going to support `detail::string`.
So, we need a new API called `get_info_impl<>` that can return `detail::strin`g.
Could you please tell me what should be the return type of `get_info`?

Reply:
> It seems you are still claiming that the return type of get_info<> is unchanged. Right?

What I'm saying is that if you define `convert_to_abi_neutral` as an identity operation under non-preview mode, then you can modify non-preview return to wrap its return value with it, which would still be an NFC change. Once the change is done, both branches of preview would have the exact same return statement.

Reply:
Something like this, yes.